### PR TITLE
Feature/task 425 implement compute quotient for bisimulation class

### PR DESCRIPTION
**Pull request by vivekramayanam**
Creating a pull request for merging ComputeQuotient Implementation

### CODE REVIEW

File: de.dlr.sc.virsat.fdir.core.test/src/de/dlr/sc/virsat/fdir/core/markov/algorithm/BisimulationTest.java [hunk 1]
```diff
@@ -249,6 +250,144 @@ public void testlabelduplicacy() {
...
+		bisimulation.computeQuotient();
+		
+		List<MarkovState> states = ma.getStates();
+		int statesCount = states.size();
+		final int requiredFinalStates = 5;
+		List<Object> stateLabelsB1 = ma.getSuccEvents(b1);
+		List<Object> stateLabelsB2 = ma.getSuccEvents(b2);
+		MarkovState stateS12 = states.contains(s1) ? s1 : s2;
+		List<Object> stateLabelsS12 = ma.getSuccEvents(stateS12);
+        List<Object> stateLabelsS3 = ma.getSuccEvents(s3);
```
Comment (SaMuellerDLR) [inline de.dlr.sc.virsat.fdir.core.test/src/de/dlr/sc/virsat/fdir/core/markov/algorithm/BisimulationTest.java:352]: Whitespacing seems off here.

File: de.dlr.sc.virsat.fdir.core/src/de/dlr/sc/virsat/fdir/core/markov/MarkovAutomaton.java [hunk 2]
```diff
@@ -66,185 +70,311 @@
...
+	 * outputs the list of events from a state to its successors
+	 * 
+	 * @param state
+	 * @return list of events of a state
+	 */
+	public List<Object> getSuccEvents(Object state) {
+		List<Object> stateLabels = new ArrayList<Object>();
+		List<MarkovTransition<S>> stateOutgoingTransitions = getSuccTransitions(state);
+		for (MarkovTransition<S> stateOutgoingTransition : stateOutgoingTransitions) {
+
```
Comment (SaMuellerDLR) [inline de.dlr.sc.virsat.fdir.core/src/de/dlr/sc/virsat/fdir/core/markov/MarkovAutomaton.java:90]: remove blank line

File: de.dlr.sc.virsat.fdir.core/src/de/dlr/sc/virsat/fdir/core/markov/MarkovAutomaton.java [hunk 3]
```diff
@@ -66,185 +70,311 @@
...
+			stateLabels.add(stateOutgoingTransition.getEvent());
+
```
Comment (SaMuellerDLR) [inline de.dlr.sc.virsat.fdir.core/src/de/dlr/sc/virsat/fdir/core/markov/MarkovAutomaton.java:92]: remove blank line

File: de.dlr.sc.virsat.fdir.core/src/de/dlr/sc/virsat/fdir/core/markov/MarkovAutomaton.java [hunk 4]
```diff
@@ -66,185 +70,311 @@
...
+	 */
+	public void removeInvalidTransitions(S state) {
+		List<MarkovTransition<S>> succTransitions = getSuccTransitions(state);
+
+		for (int i = succTransitions.size() - 1; i >= 0; i--) {
+			MarkovTransition<S> outgoingTransition = succTransitions.get(i);
+
+			MarkovState toState = outgoingTransition.getTo();
+			if (!getStates().contains(toState)) {
+
```
Comment (SaMuellerDLR) [inline de.dlr.sc.virsat.fdir.core/src/de/dlr/sc/virsat/fdir/core/markov/MarkovAutomaton.java:276]: remove blank line

File: de.dlr.sc.virsat.fdir.core/src/de/dlr/sc/virsat/fdir/core/markov/MarkovAutomaton.java [hunk 5]
```diff
@@ -66,185 +70,311 @@
...
+		List<MarkovTransition<S>> predTransitions = getPredTransitions(state);
+
+		for (int i = predTransitions.size() - 1; i >= 0; i--) {
+			MarkovTransition<S> incomingTransition = predTransitions.get(i);
+
+			MarkovState fromState = incomingTransition.getFrom();
+			if (!getStates().contains(fromState)) {
+				mapStateToPredTransitions.get(state).remove(incomingTransition);
+
```
Comment (SaMuellerDLR) [inline de.dlr.sc.virsat.fdir.core/src/de/dlr/sc/virsat/fdir/core/markov/MarkovAutomaton.java:289]: remove blank line

File: de.dlr.sc.virsat.fdir.core/src/de/dlr/sc/virsat/fdir/core/markov/algorithm/Bisimulation.java [hunk 6]
```diff
@@ -272,6 +273,59 @@ private void applyRefinement(Set<Set<MarkovState>> blocks, Set<Set<MarkovState>>
...
+			for (int i = stateOutgoingTransitions.size() - 1; i >= 0; i--) {
+				MarkovTransition<MarkovState> stateOutgoingTransition = stateOutgoingTransitions.get(i);
+
+				MarkovState toState = stateOutgoingTransition.getTo();
+				Object stateEvent = stateOutgoingTransition.getEvent();
+				MarkovState blockRepresentative = mapStateToBlock.get(toState).iterator().next();
+
+				if (toState != blockRepresentative) {
+					if (blockRepresentative != state) {
+
```
Comment (SaMuellerDLR) [inline de.dlr.sc.virsat.fdir.core/src/de/dlr/sc/virsat/fdir/core/markov/algorithm/Bisimulation.java:299]: remove blank line

**Review by SaMuellerDLR — CHANGES_REQUESTED** [review]
Please clean up whitespaces!

File: de.dlr.sc.virsat.fdir.core/src/de/dlr/sc/virsat/fdir/core/markov/algorithm/Bisimulation.java (hunk 6 above)
Comment (vivekramayanam) [reply de.dlr.sc.virsat.fdir.core/src/de/dlr/sc/virsat/fdir/core/markov/algorithm/Bisimulation.java]: removed

**Review by SaMuellerDLR — APPROVED** [review]
Approved!

**Pull request merged**